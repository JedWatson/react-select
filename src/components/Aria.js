// @flow
import React from 'react';

import { Div, SROnly } from '../primitives';

type StatusProps = { availableResults: number };
export const AriaStatus = ({ availableResults }: StatusProps) => (
  <SROnly aria-atomic="true" aria-live="polite" role="status">
    {availableResults} results are available.
  </SROnly>
);
