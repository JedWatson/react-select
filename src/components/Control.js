// @flow
import React from 'react';

import { Div } from '../primitives';
import { borderRadius, colors, spacing } from '../theme';

type FocusType = { isDisabled: boolean, isFocused: boolean };

const Control = ({ isDisabled, isFocused, ...props }: FocusType) => (
  <Div
    css={{
      alignItems: 'center',
      backgroundColor: isDisabled
        ? colors.neutral5
        : isFocused ? colors.neutral0 : colors.neutral2,
      borderColor: isDisabled
        ? colors.neutral10
        : isFocused ? colors.primary : colors.neutral20,
      borderRadius: borderRadius,
      borderStyle: 'solid',
      borderWidth: 1,
      boxShadow: isFocused ? `0 0 0 1px ${colors.primary}` : null,
      cursor: 'default',
      display: 'flex ',
      flexWrap: 'wrap',
      justifyContent: 'space-between',
      minHeight: spacing.controlHeight,
      outline: '0 !important',
      position: 'relative',
      transition:
        'background-color 100ms, border-color 100ms, box-shadow 100ms',

      '&:hover': {
        borderColor: isFocused ? colors.primary : colors.neutral30,
      },
    }}
    {...props}
  />
);

export default Control;
