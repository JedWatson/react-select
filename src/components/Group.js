// @flow
import React, { Children, cloneElement, type Node } from 'react';

import { paddingHorizontal, paddingVertical } from '../mixins';
import { Li, Ul, Strong } from '../primitives';
import { spacing } from '../theme';

type Props = {
  children: Node,
  label: Node,
};

const Group = ({ children, label, ...props }: Props) => {
  const cloneProps = { withinGroup: true, ...props };

  return (
    <Li
      aria-label={label}
      aria-expanded="true"
      role="group"
      css={paddingVertical(spacing.baseUnit * 2)}
      {...props}
    >
      <GroupHeading>{label}</GroupHeading>
      <Ul>{Children.map(children, k => cloneElement(k, cloneProps))}</Ul>
    </Li>
  );
};

const GroupHeading = props => (
  <Strong
    css={{
      color: '#999',
      cursor: 'default',
      display: 'block',
      fontSize: '75%',
      fontWeight: '500',
      marginBottom: '0.25em',
      ...paddingHorizontal(spacing.baseUnit * 3),
      textTransform: 'uppercase',
    }}
    {...props}
  />
);

export default Group;
