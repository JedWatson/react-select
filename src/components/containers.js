// @flow
import React, { Component, type Node, type ElementRef } from 'react';

import { Div } from '../primitives';
import { spacing } from '../theme';
import type { CommonProps, KeyboardEventHandler } from '../types';

// ==============================
// Root Container
// ==============================

type ContainerState = {
  /** Whether the select is disabled. */
  isDisabled: boolean,
  /** Whether the text in the select is indented from right to left. */
  isRtl: boolean,
};

export type ContainerProps = CommonProps &
  ContainerState & {
    /** The children to be rendered. */
    children: Node,
    /** Inner props to be passed down to the container. */
    innerProps: { onKeyDown: KeyboardEventHandler },
  };
export const containerCSS = ({ isDisabled, isRtl }: ContainerState) => ({
  direction: isRtl ? 'rtl' : null,
  pointerEvents: isDisabled ? 'none' : null, // cancel mouse events when disabled
  position: 'relative',
});
export const SelectContainer = (props: ContainerProps) => {
  const { children, cx, getStyles, innerProps, isDisabled, isRtl } = props;
  return (
    <Div
      css={getStyles('container', props)}
      className={cx('', { isDisabled, isRtl })}
      {...innerProps}
    >
      {children}
    </Div>
  );
};

// ==============================
// Value Container
// ==============================

export type ValueContainerProps = CommonProps & {
  /** Set when the value container should hold multiple values. This is important for styling. */
  isMulti: boolean,
  /** Whether the value container currently holds a value. */
  hasValue: boolean,
  /** Whether there should be a maximum height to the container */
  maxHeight: number,
  /** The children to be rendered. */
  children: Node,
};
export const valueContainerCSS = ({ maxHeight }: ValueContainerProps) => ({
  alignItems: 'center',
  display: 'flex',
  flex: 1,
  flexWrap: 'wrap',
  maxHeight: maxHeight, // max-height allows scroll when multi
  overflowY: 'auto',
  padding: `${spacing.baseUnit / 2}px ${spacing.baseUnit * 2}px`,
  WebkitOverflowScrolling: 'touch',
  position: 'relative',
});
export class ValueContainer extends Component<ValueContainerProps> {
  shouldScrollBottom: boolean = false;
  node: HTMLElement;
  componentWillUpdate() {
    if (!this.props.isMulti) return;

    // scroll only if the user was already at the bottom
    const total = this.node.scrollTop + this.node.offsetHeight;
    this.shouldScrollBottom = total === this.node.scrollHeight;
  }
  componentDidUpdate() {
    if (!this.props.isMulti) return;

    // ensure we're showing items being added by forcing scroll to the bottom
    if (this.shouldScrollBottom && this.node) {
      this.node.scrollTop = this.node.scrollHeight;
    }
  }
  getScrollContainer = (ref: ElementRef<*>) => {
    this.node = ref;
  };
  render() {
    const { children, cx, isMulti, getStyles, hasValue } = this.props;

    return (
      <Div
        innerRef={isMulti ? this.getScrollContainer : undefined}
        className={cx('value-container', { isMulti, hasValue })}
        css={getStyles('valueContainer', this.props)}
      >
        {children}
      </Div>
    );
  }
}

// ==============================
// Indicator Container
// ==============================

type IndicatorsState = {
  /** Whether the text should be rendered right to left. */
  isRtl: boolean,
};

export type IndicatorContainerProps = CommonProps &
  IndicatorsState & {
    /** The children to be rendered. */
    children: Node,
  };

export const indicatorsContainerCSS = () => ({
  alignItems: 'center',
  alignSelf: 'stretch',
  display: 'flex',
  flexShrink: 0,
});
export const IndicatorsContainer = (props: IndicatorContainerProps) => {
  const { children, cx, getStyles } = props;

  return (
    <Div
      className={cx('indicators')}
      css={getStyles('indicatorsContainer', props)}
    >
      {children}
    </Div>
  );
};
