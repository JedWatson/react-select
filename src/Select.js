// @flow
// @jsx glam
import React, { Component, type Element, type ElementRef } from 'react';
import glam from 'glam';

import { defaultComponents, type SelectComponents } from './components/index';
import { AriaStatus } from './components/Aria';
import { defaultFormatters, type Formatters } from './formatters';

import type {
  ActionMeta,
  FocusDirection,
  OptionType,
  OptionsType,
  ValueType,
} from './types';

/*
// TODO: make sure these are implemented comprehensively
type Customisations = {
	formatters: {
		optionLabel: OptionType => Node,
		valueLabel: OptionType => Node,
		formData: ValueType => string | Array<string>,
	},
	logic: {
		filterOptions: OptionsType => OptionsType,
		isSelected: OptionType => boolean,
		isDisabled: OptionType => boolean,
	},
};
*/

type Props = {
  clearable: boolean,
  closeMenuOnSelect: boolean,
  escapeClearsValue: boolean,
  backspaceRemovesValue: boolean,
  components: SelectComponents,
  deleteRemovesValue: boolean,
  disabledKey: string,
  formatters: Formatters,
  hideSelectedOptions: boolean,
  instanceId?: number | string,
  isDisabled: boolean,
  isMulti: boolean,
  label: string,
  maxMenuHeight: number,
  maxValueHeight: number,
  onChange: (ValueType, ActionMeta) => void,
  onKeyDown: (SyntheticKeyboardEvent<HTMLElement>) => void,
  options: OptionsType,
  placeholder?: string,
  tabSelectsValue: boolean,
  value: ValueType,
};

const defaultProps = {
  clearable: true,
  closeMenuOnSelect: true,
  escapeClearsValue: false,
  backspaceRemovesValue: true,
  deleteRemovesValue: true,
  disabledKey: 'disabled',
  hideSelectedOptions: true,
  isDisabled: false,
  isMulti: false,
  maxMenuHeight: 300,
  maxValueHeight: 100,
  options: [],
  placeholder: 'Select...',
  tabSelectsValue: true,
};

type MenuOptions = {
  render: Array<OptionType>,
  focusable: Array<OptionType>,
};

type State = {
  inputValue: string,
  inputIsHidden: boolean,
  isFocused: boolean,
  focusedOption: OptionType | null,
  menuIsOpen: boolean,
  menuOptions: MenuOptions,
  selectValue: OptionsType,
};

type ElRef = ElementRef<*>;

let instanceId = 1;

const inputStyle = {
  background: 'transparent',
  border: 0,
  fontSize: 'inherit',
  outline: 0,
};

const filterOption = (optionLabel: string, inputValue: string) => {
  return optionLabel.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
};

const cleanValue = (value: ValueType): OptionsType => {
  if (Array.isArray(value)) return value.filter(Boolean);
  if (typeof value === 'object' && value !== null) return [value];
  return [];
};

const toKey = (str: string): string => {
  return str.replace(/\W/g, '-');
};

// TODO @jedwatson `findIndex` not used. Safe to remove?
// const findIndex = (arr: Array<*>, match: any): number => {
// 	let index = 0;
// 	while (index < arr.length) {
// 		if (arr[index] === match) return index;
// 	}
// 	return -1;
// };

const scrollIntoView = (menuEl: HTMLElement, focusedEl: HTMLElement): void => {
  // TODO: Is there a way to overscroll to group headings?
  const menuRect = menuEl.getBoundingClientRect();
  const focusedRect = focusedEl.getBoundingClientRect();
  const overScroll = focusedEl.offsetHeight / 3;
  if (focusedRect.bottom + overScroll > menuRect.bottom) {
    menuEl.scrollTop = Math.min(
      focusedEl.offsetTop +
        focusedEl.clientHeight -
        menuEl.offsetHeight +
        overScroll,
      menuEl.scrollHeight
    );
  } else if (focusedRect.top - overScroll < menuRect.top) {
    menuEl.scrollTop = Math.max(focusedEl.offsetTop - overScroll, 0);
  }
};

// TODO: turn this into a prop or measure it when there's a menu ref
const PAGE_SIZE = 5;

export default class Select extends Component<Props, State> {
  static defaultProps = defaultProps;
  components: SelectComponents;
  controlRef: ElRef;
  focusedOptionRef: ?HTMLElement;
  formatters: Formatters;
  hasGroups: boolean = false;
  input: ?ElRef;
  inputHeight: ?number = 20;
  instancePrefix: string = '';
  menuRef: ?HTMLElement;
  openAfterFocus: boolean = false;
  scrollToFocusedOptionOnUpdate: boolean = false;
  state = {
    inputIsHidden: false,
    inputValue: '',
    isFocused: false,
    menuIsOpen: false,
    menuOptions: { render: [], focusable: [] },
    focusedOption: null,
    selectValue: [],
  };
  constructor(props: Props) {
    super(props);
    const { options, value } = props;
    this.components = defaultComponents(props);
    this.formatters = defaultFormatters(props);
    this.instancePrefix =
      'react-select-' + (this.props.instanceId || ++instanceId) + '-';

    const selectValue = cleanValue(value);
    const menuOptions = this.buildMenuOptions(options, selectValue);

    this.state.menuOptions = menuOptions;
    this.state.selectValue = selectValue;
  }
  componentDidMount() {
    if (this.props.autoFocus) {
      this.focus();
    }
  }
  componentWillReceiveProps(nextProps: Props) {
    const { components, options, value } = this.props;
    const { inputValue } = this.state;
    if (nextProps.components !== components) {
      this.components = defaultComponents(nextProps);
    }
    if (nextProps.value !== value || nextProps.options !== options) {
      const selectValue = cleanValue(nextProps.value);
      const menuOptions = this.buildMenuOptions(
        nextProps.options,
        selectValue,
        inputValue
      );
      const focusedOption = this.getNextFocusedOption(menuOptions.focusable);
      this.setState({ menuOptions, selectValue, focusedOption });
    }
  }
  componentDidUpdate() {
    if (
      this.menuRef &&
      this.focusedOptionRef &&
      this.scrollToFocusedOptionOnUpdate
    ) {
      scrollIntoView(this.menuRef, this.focusedOptionRef);
    }
    this.scrollToFocusedOptionOnUpdate = false;
  }
  getNextFocusedOption(options: OptionsType) {
    const { focusedOption: lastFocusedOption } = this.state;
    return lastFocusedOption && options.indexOf(lastFocusedOption) > -1
      ? lastFocusedOption
      : options[0];
  }
  buildMenuOptions(
    options: OptionsType,
    selectValue: OptionsType,
    inputValue: string = ''
  ): MenuOptions {
    const { hideSelectedOptions, isMulti } = this.props;
    const render = [];
    const focusable = [];

    const toOption = (option, i) => {
      const isSelected = this.isSelected(option, selectValue);
      if (isMulti && hideSelectedOptions && isSelected) return;
      if (!filterOption(this.getOptionLabel(option), inputValue)) return;
      if (!this.isDisabled(option)) {
        focusable.push(option);
      }
      return {
        type: 'option',
        label: this.getOptionLabel(option),
        key: `${i}-${this.getOptionValue(option)}`,
        isSelected,
        onMouseOver: () => this.onOptionHover(option),
        onClick: () => this.selectValue(option),
        data: option,
      };
    };

    options.forEach((item, itemIndex) => {
      if (item.options) {
        // TODO needs a tidier implementation
        if (!this.hasGroups) this.hasGroups = true;

        const { options: items } = item;
        const children = items.map(toOption).filter(Boolean);
        if (children.length) {
          const itemLabel = this.getOptionLabel(item);
          render.push({
            type: 'group',
            key: `${itemIndex}-${toKey(itemLabel)}`,
            label: itemLabel,
            children,
          });
        }
      } else {
        const option = toOption(item, itemIndex);
        if (option) render.push(option);
      }
    });
    return { render, focusable };
  }
  buildStateForInputValue(inputValue: string = '') {
    const { options } = this.props;
    const { selectValue } = this.state;
    const menuOptions = this.buildMenuOptions(options, selectValue, inputValue);
    const focusedOption = this.getNextFocusedOption(menuOptions.focusable);
    return { inputValue, menuOptions, focusedOption };
  }
  hasValue() {
    const { selectValue } = this.state;
    return selectValue.length > 0;
  }
  hasOptions(options?: { length: boolean }) {
    const length = options && options.length;
    const count = this.state.menuOptions.render.length;
    return length ? count : Boolean(count);
  }
  isDisabled = (option: OptionType): boolean => {
    const { disabledKey } = this.props;
    if (option[disabledKey]) return true;
    return false;
  };
  isSelected = (option: OptionType, selectValue: OptionsType): boolean => {
    if (selectValue.indexOf(option) > -1) return true;
    return selectValue.some(
      i => this.getOptionValue(i) === this.getOptionValue(option)
    );
  };
  focus() {
    if (!this.input) return;
    this.input.focus();
  }
  blurInput() {
    if (!this.input) return;
    this.input.blur();
  }
  openMenu(focusOption: 'first' | 'last' = 'first') {
    const { menuOptions, selectValue } = this.state;
    const { isMulti } = this.props;

    let openAtIndex =
      focusOption === 'first' ? 0 : menuOptions.focusable.length - 1;

    if (!isMulti) {
      const selectedIndex = menuOptions.focusable.indexOf(selectValue[0]);
      if (selectedIndex > -1) {
        openAtIndex = selectedIndex;
      }
    }

    this.scrollToFocusedOptionOnUpdate = true;
    this.setState({
      menuIsOpen: true,
      focusedOption: menuOptions.focusable[openAtIndex],
      inputIsHidden: false,
    });
  }
  focusOption(direction: FocusDirection = 'first') {
    const { focusedOption, menuOptions } = this.state;
    const options = menuOptions.focusable;
    if (!options.length) return;
    let nextFocus = 0; // handles 'first'
    const focusedIndex = focusedOption ? options.indexOf(focusedOption) : -1;
    if (direction === 'up') {
      nextFocus = focusedIndex > 0 ? focusedIndex - 1 : options.length - 1;
    } else if (direction === 'down') {
      nextFocus = (focusedIndex + 1) % options.length;
    } else if (direction === 'pageup') {
      nextFocus = focusedIndex - PAGE_SIZE;
      if (nextFocus < 0) nextFocus = 0;
    } else if (direction === 'pagedown') {
      nextFocus = focusedIndex + PAGE_SIZE;
      if (nextFocus > options.length - 1) nextFocus = options.length - 1;
    } else if (direction === 'last') {
      nextFocus = options.length - 1;
    }
    this.scrollToFocusedOptionOnUpdate = true;
    this.setState({
      focusedOption: options[nextFocus],
    });
  }
  selectValue = (newValue: OptionType) => {
    const { closeMenuOnSelect, isMulti } = this.props;
    // We update the state here because we should clear inputValue when an
    // option is selected; the onChange event fires when that's reconciled
    // otherwise the new menu items will be filtered with the old inputValue
    this.setState(
      closeMenuOnSelect
        ? {
            menuIsOpen: false,
            inputIsHidden: isMulti ? false : true,
            ...this.buildStateForInputValue(),
          }
        : this.buildStateForInputValue(),
      () => {
        const { selectValue } = this.state;
        if (isMulti) {
          if (this.isSelected(newValue, selectValue)) {
            this.props.onChange(selectValue.filter(i => i !== newValue), {
              action: 'deselect-value',
            });
          } else {
            this.props.onChange([...selectValue, newValue], {
              action: 'select-option',
            });
          }
        } else {
          this.props.onChange(newValue, { action: 'select-option' });
        }
      }
    );
  };
  removeValue = (removedValue: OptionType) => {
    const { selectValue } = this.state;
    this.props.onChange(selectValue.filter(i => i !== removedValue), {
      action: 'remove-value',
    });
    this.focus();
  };
  clearValue = () => {
    const { isMulti } = this.props;
    this.props.onChange(isMulti ? [] : null, { action: 'clear' });
  };
  popValue = () => {
    const { selectValue } = this.state;
    this.props.onChange(selectValue.slice(0, selectValue.length - 1), {
      action: 'pop-value',
    });
  };
  onControlRef = (ref: ElRef) => {
    this.controlRef = ref;
  };
  onControlMouseDown = (event: SyntheticMouseEvent<HTMLElement>) => {
    if (!this.state.isFocused) {
      this.openAfterFocus = true;
      this.focus();
    } else if (!this.state.menuIsOpen) {
      this.openMenu('first');
    } else {
      this.setState({
        menuIsOpen: false,
      });
    }
    if (event.target.tagName !== 'INPUT') {
      event.preventDefault();
    }
  };
  onKeyDown = (event: SyntheticKeyboardEvent<HTMLElement>) => {
    const {
      backspaceRemovesValue,
      clearable,
      deleteRemovesValue,
      escapeClearsValue,
      isDisabled,
      onKeyDown,
      tabSelectsValue,
    } = this.props;
    const { focusedOption, inputValue, menuIsOpen } = this.state;

    if (isDisabled) return;

    if (typeof onKeyDown === 'function') {
      onKeyDown(event);
      if (event.defaultPrevented) {
        return;
      }
    }

    switch (event.keyCode) {
      case 8: // backspace
        if (inputValue || !backspaceRemovesValue) return;
        this.popValue();
        break;
      case 46: // delete
        if (inputValue || !deleteRemovesValue) return;
        this.popValue();
        break;
      case 9: // tab
        if (
          event.shiftKey ||
          !menuIsOpen ||
          !tabSelectsValue ||
          !focusedOption
        ) {
          return;
        }
        this.selectValue(focusedOption);
        return;
      case 13: // enter
        if (menuIsOpen) {
          if (!focusedOption) return;
          this.selectValue(focusedOption);
        } else {
          this.focusOption();
        }
        break;
      case 27: // escape
        if (menuIsOpen) {
          this.setState({
            menuIsOpen: false,
            ...this.buildStateForInputValue(),
          });
        } else if (clearable && escapeClearsValue) {
          this.clearValue();
        }
        break;
      case 32: // space
        if (inputValue) {
          return;
        }
        if (!menuIsOpen) {
          this.openMenu();
          break;
        }
        if (!focusedOption) return;
        this.selectValue(focusedOption);
        break;
      case 38: // up
        if (menuIsOpen) {
          this.focusOption('up');
        } else {
          this.openMenu('last');
        }
        break;
      case 40: // down
        if (menuIsOpen) {
          this.focusOption('down');
        } else {
          this.openMenu('first');
        }
        break;
      case 33: // page up
        if (!menuIsOpen) return;
        this.focusOption('pageup');
        break;
      case 34: // page down
        if (!menuIsOpen) return;
        this.focusOption('pagedown');
        break;
      case 36: // home key
        if (!menuIsOpen) return;
        this.focusOption('first');
        break;
      case 35: // end key
        if (!menuIsOpen) return;
        this.focusOption('last');
        break;
      default:
        return;
    }
    event.preventDefault();
  };
  onInputRef = (input: ElRef) => {
    this.input = input;

    // cache the input height to use when the select is disabled
    if (input && input.input) this.inputHeight = input.input.clientHeight;
  };
  onInputChange = (event: SyntheticKeyboardEvent<HTMLInputElement>) => {
    const inputValue = event.currentTarget.value;
    this.setState({
      inputIsHidden: false,
      menuIsOpen: true,
      ...this.buildStateForInputValue(inputValue),
    });
  };
  onInputFocus = (event: SyntheticFocusEvent<HTMLInputElement>) => {
    this.setState({
      inputIsHidden: false,
      isFocused: true,
    });
    if (this.openAfterFocus) {
      this.openMenu('first');
    }
    this.openAfterFocus = false;
  };
  onInputBlur = (event: SyntheticFocusEvent<HTMLInputElement>) => {
    this.setState({
      isFocused: false,
      menuIsOpen: false,
      ...this.buildStateForInputValue(''),
    });
  };
  onMenuRef = (ref: ?HTMLElement) => {
    this.menuRef = ref;
  };
  onMenuMouseDown = (event: SyntheticMouseEvent<HTMLElement>) => {
    if (event.button !== 0) {
      return;
    }
    event.stopPropagation();
    event.preventDefault();
    this.focus();
  };
  onFocusedOptionRef = (ref: ?HTMLElement) => {
    this.focusedOptionRef = ref;
  };
  onOptionHover = (data: OptionType) => {
    this.setState({
      focusedOption: data,
    });
  };
  onDropdownIndicatorMouseDown = (event: SyntheticMouseEvent<HTMLElement>) => {
    // ignore mouse events that weren't triggered by the primary button
    if (event && event.type === 'mousedown' && event.button !== 0) {
      return;
    }
    if (this.props.isDisabled) return;
    const { isMulti } = this.props;
    const { menuIsOpen } = this.state;
    if (!this.focused) {
      this.focus();
    }
    if (menuIsOpen) {
      this.setState({
        menuIsOpen: false,
        inputIsHidden: isMulti ? false : true,
      });
    } else {
      this.openMenu();
    }
    event.preventDefault();
    event.stopPropagation();
  };
  onClearIndicatorMouseDown = (event: SyntheticMouseEvent<HTMLElement>) => {
    // ignore mouse events that weren't triggered by the primary button
    if (event && event.type === 'mousedown' && event.button !== 0) {
      return;
    }
    this.clearValue();
    event.stopPropagation();
    this.openAfterFocus = false;
    setTimeout(() => this.focus());
  };
  getElementId = (element: 'input' | 'label' | 'listbox' | 'option') => {
    return `${this.instancePrefix}-${element}`;
  };
  getActiveDescendentId = () => {
    const { focusedOption, menuIsOpen } = this.state;
    return menuIsOpen && focusedOption
      ? `${this.getElementId('option')}-${this.getOptionValue(focusedOption)}`
      : undefined;
  };
  renderInput(id: string) {
    const { isDisabled } = this.props;
    const { Input } = this.components;
    const { inputIsHidden, inputValue, menuIsOpen } = this.state;

    // maintain baseline alignment when the input is removed
    if (isDisabled) return <div style={{ height: this.inputHeight }} />;

    // aria properties makes the JSX "noisy", separated for clarity
    const ariaAttributes = {
      'aria-activedescendant': this.getActiveDescendentId(),
      'aria-autocomplete': 'list',
      'aria-expanded': menuIsOpen,
      'aria-haspopup': menuIsOpen,
      'aria-owns': menuIsOpen ? this.getElementId('listbox') : undefined,
      role: 'combobox',
    };

    return (
      <Input
        autoCapitalize="none"
        autoComplete="off"
        autoCorrect="off"
        id={id}
        innerRef={this.onInputRef}
        inputStyle={{ ...inputStyle, opacity: inputIsHidden ? 0 : 1 }}
        onBlur={this.onInputBlur}
        onChange={this.onInputChange}
        onFocus={this.onInputFocus}
        spellCheck="false"
        tabIndex="0"
        type="text"
        value={inputValue}
        {...ariaAttributes}
      />
    );
  }
  getOptionLabel(data: OptionType) {
    return this.formatters.optionLabel(data);
  }
  getOptionValue(data: OptionType) {
    return this.formatters.optionValue(data);
  }
  getValueLabel(data: OptionType) {
    return this.formatters.valueLabel(data);
  }
  renderPlaceholderOrValue() {
    const { MultiValue, SingleValue, Placeholder } = this.components;
    const { isDisabled, isMulti, placeholder } = this.props;
    const { inputValue, selectValue } = this.state;

    if (!this.hasValue()) {
      return inputValue ? null : (
        <Placeholder
          key="placeholder"
          isDisabled={isDisabled}
          isMulti={isMulti}
        >
          {placeholder}
        </Placeholder>
      );
    }
    if (isMulti) {
      return selectValue.map(opt => (
        <MultiValue
          isDisabled={isDisabled}
          label={this.getValueLabel(opt)}
          key={this.getOptionValue(opt)}
          onRemove={this.removeValue}
          data={opt}
        />
      ));
    }
    if (inputValue) return null;
    const singleValue = selectValue[0];
    return (
      <SingleValue
        children={this.getValueLabel(singleValue)}
        data={singleValue}
        isDisabled={isDisabled}
      />
    );
  }
  renderClearIndicator() {
    const { ClearIndicator } = this.components;
    const { isDisabled } = this.props;
    const { isFocused } = this.state;

    if (!ClearIndicator || isDisabled || !this.hasValue()) return null;

    return (
      <ClearIndicator
        isFocused={isFocused}
        onMouseDown={this.onClearIndicatorMouseDown}
      />
    );
  }
  renderDropdownIndicator() {
    const { DropdownIndicator } = this.components;
    if (!DropdownIndicator) return null;
    const { isFocused } = this.state;

    return (
      <DropdownIndicator
        isFocused={isFocused}
        onMouseDown={this.onDropdownIndicatorMouseDown}
      />
    );
  }

  renderMenu() {
    const { Group, Menu, MenuList, Option, NoOptions } = this.components;
    const { focusedOption, menuIsOpen, menuOptions } = this.state;
    const { isMulti, maxMenuHeight } = this.props;

    if (!menuIsOpen) return null;

    // TODO: Internal Option Type here
    const render = (option: OptionType) => {
      const id = `${this.getElementId('option')}-${this.getOptionValue(
        option.data
      )}`;
      const isFocused = focusedOption === option.data;
      return (
        <Option
          {...option}
          id={id}
          isFocused={isFocused}
          innerRef={isFocused ? this.onFocusedOptionRef : undefined}
        >
          {option.label}
        </Option>
      );
    };

    let menuUI;

    if (this.hasOptions()) {
      menuUI = menuOptions.render.map(item => {
        if (item.type === 'group') {
          const { children, type, ...group } = item;
          return (
            <Group {...group}>
              {item.children.map(option => render(option))}
            </Group>
          );
        } else if (item.type === 'option') {
          return render(item);
        }
      });
    } else {
      menuUI = <NoOptions>No options...</NoOptions>;
    }

    return (
      <Menu onMouseDown={this.onMenuMouseDown}>
        <MenuList
          aria-labelledby={this.getElementId('label')}
          id={this.getElementId('listbox')}
          innerRef={this.onMenuRef}
          isMulti={isMulti}
          maxHeight={maxMenuHeight}
          role={this.hasGroups ? 'tree' : 'listbox'}
        >
          {menuUI}
        </MenuList>
      </Menu>
    );
  }
  render() {
    const {
      Control,
      DropdownIndicator,
      IndicatorsContainer,
      Label,
      SelectContainer,
      ValueContainer,
    } = this.components;

    const { isDisabled, isMulti, label, maxValueHeight } = this.props;
    const { isFocused } = this.state;
    const inputId = this.getElementId('input');
    const labelId = this.getElementId('label');

    // TODO
    // - return React.Fragment when v16
    // - add `aria-busy` to SelectContainer when loading async
    return (
      <div>
        {label ? (
          <Label htmlFor={inputId} id={labelId}>
            {label}
          </Label>
        ) : null}
        <SelectContainer isDisabled={isDisabled} onKeyDown={this.onKeyDown}>
          <AriaStatus availableResults={this.hasOptions({ length: true })} />
          <Control
            isDisabled={isDisabled}
            isFocused={isFocused}
            onMouseDown={this.onControlMouseDown}
            innerRef={this.onControlRef}
          >
            <ValueContainer
              isMulti={isMulti}
              hasValue={this.hasValue()}
              maxHeight={maxValueHeight}
            >
              {this.renderPlaceholderOrValue()}
              {this.renderInput(inputId)}
            </ValueContainer>
            <IndicatorsContainer>
              {this.renderClearIndicator()}
              {this.renderDropdownIndicator()}
            </IndicatorsContainer>
          </Control>
          {this.renderMenu()}
        </SelectContainer>
      </div>
    );
  }
}
