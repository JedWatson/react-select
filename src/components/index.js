// @flow
import {
  IndicatorsContainer,
  SelectContainer,
  ValueContainer,
} from './containers';
import { ClearIndicator, DropdownIndicator } from './indicators';

import Control from './Control';
import Group from './Group';
import Input from './Input';
import Label from './Label';
import Menu, { MenuList, NoOptions } from './Menu';
import MultiValue from './MultiValue';
import Option from './Option';
import Placeholder from './Placeholder';
import SingleValue from './SingleValue';

export type SelectComponents = {
  ClearIndicator: typeof ClearIndicator,
  Control: typeof Control,
  DropdownIndicator: typeof DropdownIndicator,
  Group: typeof Group,
  Label: typeof Label,
  IndicatorsContainer: typeof IndicatorsContainer,
  Input: typeof Input,
  Menu: typeof Menu,
  MenuList: typeof MenuList,
  MultiValue: typeof MultiValue,
  NoOptions: typeof NoOptions,
  Option: typeof Option,
  Placeholder: typeof Placeholder,
  SelectContainer: typeof SelectContainer,
  SingleValue: typeof SingleValue,
  ValueContainer: typeof ValueContainer,
};

export const components: SelectComponents = {
  ClearIndicator: ClearIndicator,
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Group: Group,
  Label: Label,
  IndicatorsContainer: IndicatorsContainer,
  Input: Input,
  Menu: Menu,
  MenuList: MenuList,
  MultiValue: MultiValue,
  NoOptions: NoOptions,
  Option: Option,
  Placeholder: Placeholder,
  SelectContainer: SelectContainer,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer,
};

type Props = {
  components: SelectComponents,
};

export const defaultComponents = (props: Props) => ({
  ...components,
  ...props.components,
});
