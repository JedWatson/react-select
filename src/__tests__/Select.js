import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Select from '../Select';
import { components } from '../components';

const { Option, SingleValue } = components;

test('defaults', () => {
	const tree = shallow(<Select />);
	expect(toJson(tree)).toMatchSnapshot();
});

test('formatters.optionLabel', () => {
	function optionLabel({ label, value }) {
		return `${label} ${value}`;
	}
	const tree = shallow(
		<Select
			formatters={{ optionLabel }}
			options={[
				{ label: '1', value: 'one' },
				{ label: '2', value: 'two' },
				{ label: '3', value: 'three' },
			]}
		/>,
	);
	tree.setState({ menuIsOpen: true });
	const opts = tree.find(Option);
	expect(opts).toHaveLength(3);
	expect(opts.at(0).props().label).toBe('1 one');
	expect(opts.at(1).props().label).toBe('2 two');
	expect(opts.at(2).props().label).toBe('3 three');
});

test('formatters.valueLabel', () => {
	function valueLabel({ label, value }) {
		return `${label} ${value}`;
	}
	const opts = [
		{ label: '1', value: 'one' },
		{ label: '2', value: 'two' },
		{ label: '3', value: 'three' },
	];
	const tree = shallow(<Select formatters={{ valueLabel }} options={opts} />);
	tree.setState({ selectValue: [opts[1]] });
	const sval = tree.find(SingleValue).at(0);
	expect(sval.props().label).toBe('2 two');
});
