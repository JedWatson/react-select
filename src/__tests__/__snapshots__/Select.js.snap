// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
<div>
  <SelectContainer
    aria-expanded={false}
    isDisabled={false}
    onKeyDown={[Function]}
    role="combobox"
  >
    <AriaStatus
      availableResults={0}
    />
    <Control
      isDisabled={false}
      isFocused={false}
      onClick={[Function]}
    >
      <ValueContainer
        hasValue={false}
        isMulti={false}
      >
        <Placeholder>
          Select...
        </Placeholder>
        <AutosizeInput
          autoCapitalize="none"
          autoComplete="off"
          autoCorrect="off"
          id="react-select-2--input"
          injectStyles={true}
          inputStyle={
            Object {
              "background": "transparent",
              "border": 0,
              "fontSize": "inherit",
              "opacity": 100,
              "outline": 0,
            }
          }
          minWidth={1}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          role="textbox"
          spellCheck="false"
          tabIndex="0"
          value=""
        />
      </ValueContainer>
      <Component>
        <Component
          isFocused={false}
          onMouseDown={[Function]}
        />
      </Component>
    </Control>
  </SelectContainer>
</div>
`;
